accepted/0017-image-slideshow-redesign.md

```diff
@@ -0,0 +1,100 @@
+slideshow.
+
+It should be able to show all of the selected images.
+
+A minor issue that may be left out of this RFC is the way random playback is behaving. Currently it picks the 
+next image completely random. This may result in some images getting shown multiple times while others may get 
+shown seldom or not at all. The current implementation also suffers from modulo bias which means the depending 
+on the number of images, the first few images may be "favored" by the algorithm and shown more often. The 
+author's expectation for random playback is to play back the whole list once, but in random order (shuffled),
+thus each image is shown exactly once before the first repeat is seen.
+
+
+# Detailed design
+
+## Main design 
+
+* The slideshow gets a list of files the user has configured to play.
+* At the start, the list is copied to a new list, the playback queue.
+  * The reason for this is to support an improvement in the way random playback is handled (this is discussed below).
+  * It also allows to remove invalid/missing images from the queue but keeps the user's intention (play this 
+    file), so at the next opportunity the slideshow may try to handle this file again.
+* There is a cache/preload buffer that stores only the decoded images that are needed in the near-future or that
+  have been used recently (to handle "go back an image").
+* The video tick function fetches images from the cache.
+  * If at a tick an images had to be shown but it wasn't available in the cache yet, the transition is deferred 
+    and tried again at the next tick. This should happen seldomly, for example right at the start when the cache
+    was not yet able to load the current image.
+* A dedicated thread is responsible for managing the cache/preload content: images are loaded as required 
+  according to the current position in the queue and entries that are no longer needed are evicted, freeing
+  their resources.
+  * This thread is waiting for an event so that it's sleeping when there is no work to do and woken up once the
+    index into the playback queue is advanced.
+  * First, it ensures that the image for the current playback position is loaded.
+  * Then, it ensures that the next few images (subject to compile-time constant or setting) are loaded.
+  * Then, it ensures that the previous few images are loaded to support going backwards.
+  * Lastly, it evicts all other images from the cache, thus unloading their resources if there are no more 
+    references (for example, from a transition).
+* Increasing/decreasing the index into the queue signals the event that wakes up the caching/preloading thread
+  to do its work.
+
+
+## Random playback
+
+* At the start of playback, the file queue is shuffled.
+* The index then advances in the queue as it would if no randomization was requested.
+* If the end of the queue is reached, it is re-shuffled.
```

Comment:
I feel like random shouldn't need to do this.  When randomize is used, it just chooses a random file of the bunch to put at the end of the current cache, and it really should be random (minus the previous file).  I feel like a shuffling mechanism shouldn't be necessary at all.